trace-utils/src/send_data.rs

```diff
@@ -5,83 +5,171 @@ use anyhow::{anyhow, Context};
+        match self {
+            RequestError::Timeout => write!(f, "Connection timed out"),
+            RequestError::Network => write!(f, "Network error"),
+            RequestError::Build => write!(f, "Request failed due to invalid property"),
+        }
+    }
 }
 
+impl std::error::Error for RequestError {}
+
+pub enum RequestResult {
+    // Holds information from a succesful request.
+    Success((Response<Body>, Attempts, BytesSent, ChunksSent)),
+    // Treats HTTP errors.
+    Error((Response<Body>, Attempts, ChunksDropped)),
+    // Treats timeout errors originated in the transport layer.
+    TimeoutError((Attempts, ChunksDropped)),
+    // Treats errors coming from networking.
+    NetworkError((Attempts, ChunksDropped)),
+    // Treats errors coming from building the request
+    BuildError((Attempts, ChunksDropped)),
+}
+
+#[derive(Debug)]
 pub struct SendDataResult {
+    // Keeps track of the last request result.
     pub last_result: anyhow::Result<Response<Body>>,
+    // Count metric for 'trace_api.requests'.
     pub requests_count: u64,
+    // Count metric for 'trace_api.responses'. Each key maps  a different HTTP status code.
     pub responses_count_per_code: HashMap<u16, u64>,
+    // Count metric for 'trace_api.errors' (type: timeout).
     pub errors_timeout: u64,
+    // Count metric for 'trace_api.errors' (type: network).
     pub errors_network: u64,
+    // Count metric for 'trace_api.errors' (type: status_code).
     pub errors_status_code: u64,
+    // Count metric for 'trace_api.bytes'
+    pub bytes_sent: u64,
+    // Count metric for 'trace_chunk_sent'
+    pub chunks_sent: u64,
+    // Count metric for 'trace_chunks_dropped'
+    pub chunks_dropped: u64,
 }
 
-impl SendDataResult {
-    fn new() -> SendDataResult {
+impl Default for SendDataResult {
+    fn default() -> Self {
         SendDataResult {
             last_result: Err(anyhow!("No requests sent")),
             requests_count: 0,
             responses_count_per_code: Default::default(),
             errors_timeout: 0,
             errors_network: 0,
             errors_status_code: 0,
+            bytes_sent: 0,
+            chunks_sent: 0,
+            chunks_dropped: 0,
         }
     }
+}
 
-    async fn update(
-        &mut self,
-        res: Result<Response<Body>, SendRequestError>,
-        expected_status: StatusCode,
-    ) {
-        self.requests_count += 1;
+impl SendDataResult {
+    ///
+    /// Updates `SendDataResult` internal information with the request's result information.
+    ///
+    /// # Arguments
+    ///
+    /// * `res` - Request result.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use datadog_trace_utils::send_data::RequestResult;
+    /// use datadog_trace_utils::trace_utils::SendDataResult;
+    ///
+    /// let result = RequestResult::NetworkError((1, 0));
+    /// let mut data_result = SendDataResult::default();
+    /// data_result.update(result);
```

Comment:
This example is a bit misleading since the `.await` is missing from the call to update. You could put the example code within an async function and add the miri ignore so it compiles and doc tests passs

/// #[cfg_attr(miri, ignore)]
/// async fn test_with_poll() {
///    ...
///}